lib/interceptor/decompress.js

```diff
@@ -0,0 +1,131 @@
+'use strict'
+
+const util = require('../core/util')
+const DecoratorHandler = require('../handler/decorator-handler')
+const zlib = require('node:zlib')
+const { pipeline, finished } = require('node:stream')
+const { createInflate } = require('../web/fetch/util')
+
+const nullBodyStatus = [101, 204, 205, 304]
+
+class DecompressHandler extends DecoratorHandler {
+  #handler
+  #opts
+
+  #inputStream = null
+  #trailers = null
+
+  constructor (opts, handler) {
+    super(handler)
+    this.#handler = handler
+    this.#opts = opts
+  }
+
+  onConnect (abort) {
+    this.#inputStream = null
+    this.#trailers = null
+    return this.#handler.onConnect(abort)
+  }
+
+  #onDecompressStreamFinished (err) {
+    if (err) {
+      this.#handler.onError(err)
+    } else {
+      this.#handler.onComplete(this.#trailers)
+    }
+  }
+
+  onHeaders (statusCode, rawHeaders, resume, statusMessage) {
+    const parsedHeaders = util.parseHeaders(rawHeaders)
+    const contentEncoding = parsedHeaders['content-encoding']
+    const requestEncodings = contentEncoding ? contentEncoding.split(',').map(e => e.trim().toLowerCase()) : []
+
+    const { method } = this.#opts
+
+    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding
+    if (requestEncodings.length !== 0 && method !== 'HEAD' && method !== 'CONNECT' && !nullBodyStatus.includes(statusCode)) {
+      const decoders = []
+      for (let i = 0; i < requestEncodings.length; ++i) {
```

Comment:
```suggestion
      for (let i = requestEncodings.length - 1; i >=0; --i) {
```

Reply:
The other way round, right?

Reply:
Yeah, I made the same changes #3343.

Reply:
Thanks, I'll come back to this PR this week